[BOLT] Extend Inliner to work on functions with Pointer Authentication

bgergely0 opened the pull request:
The inliner uses DirectSP to check if a function has instructions that
modify the SP. Exceptions are stack Push and Pop instructions.

We can also allow pointer signing and authentication instructions.

The inliner removes the Return instructions from the inlined functions.
If it is a fused pointer-authentication-and-return (e.g. RETAA), we have
to generate a new authentication instruction.

bgergely0 committed:
[BOLT] Extend Inliner to work on functions with Pointer Autentication

The inliner uses DirectSP to check if a function has instructions that
modify the SP. Exceptions are stack Push and Pop instructions.

We can also allow pointer signing and authentication instructions.

The inliner removes the Return instructions from the inlined functions.
If it is a fused pointer-authentication-and-return (e.g. RETAA), we have
to generate a new authentication instruction in place of the Return.

bgergely0 commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/162458?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#162458** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/162458?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/162458?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#162251** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/162251?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

peterwaller-arm commented on bolt/lib/Target/AArch64/AArch64MCPlusBuilder.cpp:
```diff
@@ -266,6 +266,35 @@ class AArch64MCPlusBuilder : public MCPlusBuilder {
+      break;
+    case AArch64::RETABSPPCr:
+      Auth.setOpcode(AArch64::AUTIBSPPCr);
+      break;
```
The *PPCi, *PPCr variants take operands that need to be set, and are untested.

peterwaller-arm reviewed (commented):
Thanks for the patch. A couple of problems spotted - I'm thinking it would be good to test more of the variants you intend to function per the opcodes in createMatchingAuth.

bgergely0 commented on bolt/lib/Passes/Inliner.cpp:
```diff
@@ -338,6 +345,17 @@ Inliner::inlineCall(BinaryBasicBlock &CallerBB,
+        BC.MIB->createMatchingAuth(Inst, Auth);
+        InsertII =
+            std::next(InlinedBB->insertInstruction(InsertII, std::move(Auth)));
+      }
```
Thanks for the catch btw, it does cause problems.

bgergely0 replied:
should be fixed in the new commit

bgergely0 committed:
[BOLT] Fix when inlining into a context with a tailcall

When inlining to a call site with a tailcall, the return in the inlined
block does not get removed. Because of this, we don't have to generate
the matching authentication.
Add test for this case.

bgergely0 committed:
[BOLT] Fix: copy operands of MCInst in createMatchingAuth

- some PAuthAndRet variants need operands, so we need to copy them from
  the to-be-removed MCInst to the new one
- remove extra assertion
- add unittest about inlining an Armv9.5-A PAuthAndRet variant (with the
  operand copy).

bgergely0 committed:
[BOLT] Add unittest for inliner using retaasppc <label>

peterwaller-arm reviewed (approved):
Just one more tweak on the comment but otherwise LGTM.

bgergely0 committed:
Fix comment
